clusterloader2/testing/l4ilb/config-nodepool-upgrade.yaml

```diff
@@ -0,0 +1,124 @@
+# Timer measurement is used to measure the time between the start of nodepool upgrade and when
+# the ILB services are running and reachable.
+# ServiceCreationLatency measurement is used only for determining the moment when all of the
+# ILBs are reachable. Since it reports time based on the initial creation of the service, and
+# not from the nodepool upgrade, the measured time will not be shown.
+
+# Constants
+{{$LB_REPLICAS_PER_NS := DefaultParam .CL2_LB_REPLICAS_PER_NS 5}}
+{{$LB_BACKEND_SIZE := DefaultParam .CL2_LB_BACKEND_SIZE 300}}
+{{$EXTERNAL_TRAFFIC_POLICY := DefaultParam .CL2_EXTERNAL_TRAFFIC_POLICY "Cluster"}}
+{{$ilbWaitTimeout := DefaultParam .CL2_ILB_WAIT_TIMEOUT "10m"}}
+{{$ilbQPS := DefaultParam .CL2_ILB_TEST_QPS 20}}
+{{$NODEPOOL_UPGRADE_LABEL := "ilb-np-upgrade"}}
+{{$namespaces := 1}}
+
+# Command to be executed
+{{$NODEPOOL_UPGRADE_COMMAND := DefaultParam .CL2_EXEC_COMMAND nil}}
+
+name: l4ilb-nodepool-upgrade
+namespace:
+  number: {{$namespaces}}
+tuningSets:
+- name: ILBConstantQPS
+  qpsLoad:
+    qps: {{$ilbQPS}}
+steps:
+- name: Start measurement for running pods
+  measurements:
+  - Identifier: WaitForRunningDeployments
+    Method: WaitForControlledPodsRunning
+    Params:
+      action: start
+      apiVersion: apps/v1
+      kind: Deployment
+      labelSelector: group = ilb-load
+      operationTimeout: 15m
+- name: Creating ILBs
+  phases:
+  - namespaceRange:
+      min: 1
+      max: {{$namespaces}}
+    replicasPerNamespace: {{$LB_REPLICAS_PER_NS}}
+    tuningSet: ILBConstantQPS
+    objectBundle:
+    - basename: ilb-npu-service
+      objectTemplatePath: service.yaml
+      templateFillMap:
+        DeploymentBaseName: ilb-npu-dep
+        ExternalTrafficPolicy: {{$EXTERNAL_TRAFFIC_POLICY}}
+        ILBSizeLabel: {{$NODEPOOL_UPGRADE_LABEL}}
+    - basename: ilb-npu-dep
+      objectTemplatePath: dep.yaml
+      templateFillMap:
+        NumReplicas: {{$LB_BACKEND_SIZE}}
+- name: Waiting for objects creation to be completed
+  measurements:
+  - Identifier: WaitForRunningDeployments
+    Method: WaitForControlledPodsRunning
+    Params:
+      action: gather
+{{if $NODEPOOL_UPGRADE_COMMAND}}
+- name: Starting measurements
+  measurements:
+  - Identifier: ServiceCreationLatency
+    Method: ServiceCreationLatency
+    Params:
+      action: start
+      waitTimeout: {{$ilbWaitTimeout}}
+      labelSelector: size = {{$NODEPOOL_UPGRADE_LABEL}}
+- name: Starting ILB recovery timer
+  measurements:
+  - Identifier: ILbRecoveryTimer
+    Method: Timer
+    Params:
+      action: start
+      label: {{$NODEPOOL_UPGRADE_LABEL}}
+- name: Execute command for nodepool upgrade
+  measurements:
+  - Identifier: ExecNodepoolUpgrade
+    Method: Exec
+    Params:
+      command:
+      {{range $NODEPOOL_UPGRADE_COMMAND}}
+      - {{.}}
+      {{end}}
+- name: Waiting for ILBs to be reachable after nodepool upgrade
+  measurements:
+  - Identifier: ServiceCreationLatency
```

Comment:
This is good for the first iteration, but ideally we want to measure 1) latency from outside the cluster (on a different GCE VM) 2) A count of how many requests succeeded and how many failed. 

Usually we don't see complete blackholing, we see intermittent failures. So a failure count will give a better idea than the latency (which might be ~0 if the first request succeeds).

Reply:
Regarding different GCE VM - Can we measure from a CL2 container? CL2 is running in another project and this is something similar how we define SLOs. Externally visible and measured from the client's perspective.

Reply:
What is a CL2 container? Is it a pod running in a different cluster in a different project?

For internal LB, clients need to be part of the same VPC. If we can use shared VPC to achieve that, then it could work.

Reply:
CL2 - container running ClusterLoader2. It lives in different project, and adding this peering would be problematic.